Implement new bedrock listener abstraction and re-plat Kestrel on top

davidfowl opened the pull request:
This is a massive set of changes to Kestrel to remove the existing pubternal transport layer and implement a public facing API for listeners and clients, see the details here https://github.com/aspnet/AspNetCore/issues/10308.

This change only has the server side pieces of the story as I don't want to add the client APIs without having ported SignalR to use them. Here are the highlights:

- Transport.Abstractions is empty (will be removed in a separate PR as it requires removing it from a *ton* of places)
- `TransportConnection` has been moved to Connection.Abstractions (we can decide if we need to consolidate with `DefaultConnectionContext` in a later PR)
- Added `FileHandleEndPoint` which allows binding to a file handle (could be a pipe or tcp handle)
- ListenOptions has been gutted for most pubternal API and returns various types of binding information . The source of truth is the `EndPoint` instance.
- Cleaned up a bunch of libuv tests decoupling them from Kestrel.Core

## Breaking Changes
- Removing pubternal API is itself a breaking change but one that we already planned to do.
- We've removed the ability to set the scheduling mode on Kestrel, this will be a transport knob instead (https://github.com/aspnet/AspNetCore/pull/10321/commits/77d3c27c468d69e2bda0195b2a64d68a921bf783) cc @sebastienros (I know our benchmarks support this)
- DisposeAsync was added to ConnectionContext (it will implement IAsyncDisposable in a later change, it supports netstandard 2.0 so that requires a bit more work) cc @anurse @BrennanConroy as this affects SignalR (https://github.com/aspnet/AspNetCore/pull/10321/commits/4977f1b5f237a84a85ca0e0d4cd4eb7f85cd3e51)
- NoDelay is a noop, on `ListenOptions`, we should consider deprecating it. This has been moved to each of the transports. One major difference though is that it's no longer localized per endpoint but is global. We'd need a derived `EndPoint` type (or maybe extend `IPEndPoint`) to store both the socket options and the binding information.

@halter73 this is ready for review now.

Fixes #10308 
Fixes #4752

davidfowl committed:
Spike Kestrel with new transport abstraction.

davidfowl committed:
Libuv works

davidfowl committed:
Removed features

davidfowl commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnectionListener.cs:
```diff
@@ -119,9 +123,39 @@ public async Task BindAsync()
+
+            while (true)
+            {
+                (LibuvConnection connection, int slot) = await await Task.WhenAny(slots);
```
But the double await look so cool.

gfoidl replied:
Agree 😄

davidfowl commented on src/Servers/Kestrel/Core/src/ListenOptions.cs:
```diff
@@ -90,6 +90,8 @@ public FileHandleType HandleType
             }
         }
 
+        public EndPoint Endpoint => IPEndPoint;
```
There are more breaks that will need to be made, I'm still feeling it out.

davidfowl commented on src/Servers/Connections.Abstractions/src/IConnectionListener.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        ValueTask<ConnectionContext> AcceptAsync();
+
+        ValueTask DisposeAsync();
```
Yes I just hacked it because this is ns2.0.

gfoidl replied:
Ahh...thx.

ReubenBond commented on src/Servers/Kestrel/Core/src/Internal/ConnectionDispatcher.cs:
```diff
@@ -28,23 +28,45 @@ public ConnectionDispatcher(ServiceContext serviceContext, ConnectionDelegate co
+                        // that we have access to a couple of methods that only exists on TransportConnection
+                        // (specifically RequestClose, TickHeartbeat and CompleteAsync. This dependency needs to be removed or
+                        // the return value of AcceptAsync needs to change.
 
```
Changing the return value & adding a `CancellationToken` to `AcceptAsync` works for me. In-band cancellation is ideal (i.e, cancellation via return value, not exception).

davidfowl replied:
Then we can make null signify no more connections accept. I have another idea to try for the return value

ReubenBond commented on src/Servers/Connections.Abstractions/src/ConnectionContext.cs:
```diff
@@ -18,6 +20,9 @@ public abstract class ConnectionContext
 
+        public abstract EndPoint LocalEndpoint { get; set; }
+        public abstract EndPoint RemoteEndpoint { get; set; }
+
```
If we have the Endpoints here (lgtm) then it seems we don't need `IConnectionEndPointFeature` - should we axe it?

Tratcher replied:
The features are still useful upstack where ConnectionContext isn't available.

davidfowl replied:
Generally, the ConnectionContext isn't the source of truth, features are and the implementation should be using the feature. This enables decoration of features by middleware without having to make a derived ConnectionContext

davidfowl committed:
Added EndPoint to IConnectionListener

davidfowl committed:
Remove null check

davidfowl committed:
Null means no more connections

davidfowl committed:
More cleanup
- Added NoDelay to transport options
- Added FileHandleEndPoint to represent listening to a file handle

davidfowl committed:
Endpoint -> EndPoint

davidfowl committed:
Generate the connection id in kestrel

DAllanCarr commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/ListenerContext.cs:
```diff
@@ -3,39 +3,52 @@
+            switch (EndPoint)
             {
-                case ListenType.IPEndPoint:
+                case IPEndPoint ip:
```
Discards are ugly!

davidfowl committed:
Implement shutdown in the socket transport

davidfowl commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs:
```diff
@@ -101,8 +121,22 @@ public override void Abort(ConnectionAbortedException abortReason)
+            {
+                // TODO: Make this timeout configurable, this gives the task time to flush the data to the socket
+                // before timing out
+                var result = await Task.WhenAny(_task, Task.Delay(5000));
```
This doesn't *have* to timeout, the previous logic just disposed and assumed timeout logic would be handled by the upper application layers. It could just await the task without aborting.

davidfowl commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -0,0 +1,244 @@
+                var spin = new SpinWait();
+                while (_pendingAccepts > 0)
+                {
+                    spin.SpinOnce();
```
This could Task.Delay instead but it should be very fast.

davidfowl committed:
Progress towards removing the task

davidfowl committed:
Remove the KestrelConnection cast
- Moved feature implementations into KestrelConnection out of TransportConnection

davidfowl committed:
Remove IHttpConnectionFeature

davidfowl committed:
Remove comment

davidfowl committed:
Fix in memory transport

davidfowl committed:
Make more tests compile

davidfowl committed:
Use the memory pool

davidfowl committed:
Move graceful shutdown back to the caller

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelConnection.cs:
```diff
@@ -1,25 +1,158 @@
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
 {
-    internal class KestrelConnection
+    internal class KestrelConnection : IConnectionHeartbeatFeature, IConnectionCompleteFeature, IConnectionLifetimeNotificationFeature
```
@halter73 check this out. This basically moves the logic that Kestrel needs to trigger into kestrel itself instead of being in the transport layer.

davidfowl committed:
More clean up
- Reduce the amount of breaking changes by making the new members virtual
- Removed IConnectionFactory for as part of this PR to focus on the server first.
- Updated the SignalR APIs to override DisposeAsync (this is breaking)

davidfowl committed:
Revert making LibuvTransportFactory public

davidfowl committed:
Minor clean up

davidfowl committed:
Shifting and deleting things

davidfowl committed:
Make more tests compile

davidfowl committed:
Make it all compile

davidfowl committed:
Use the scheduler in the InMemoryTransport

davidfowl committed:
Use options

davidfowl committed:
Fixed port binding

davidfowl committed:
Made ListenOptinos.IPEndPoint readonly

davidfowl committed:
Fix libuv binding

davidfowl committed:
Make IPEndPoint work on ListenOptions

davidfowl committed:
More tests passing

davidfowl commented on src/Servers/Kestrel/Core/test/ConnectionDispatcherTests.cs:
```diff
@@ -50,28 +51,6 @@ public void OnConnectionCreatesLogScopeWithConnectionId()
         }
 
-        [Fact]
-        public async Task OnConnectionCompletesTransportPipesAfterReturning()
```
The transport does this now.

davidfowl committed:
EndPoint is the source of truth for binding

davidfowl committed:
Fixed test

davidfowl committed:
Minor clean up

davidfowl committed:
Fixed cloning listen options

davidfowl committed:
Wait for the accept loop to end

davidfowl committed:
Clean up ConnectionManager

davidfowl committed:
PR feedback

davidfowl committed:
Fixed timeout tests

davidfowl committed:
Made more libuv tests pass
- Decouple LibuvThread from LibuvTransport

davidfowl committed:
Fixed one more test

davidfowl committed:
Wait for connection close to fire before disposing the token

davidfowl committed:
Make last libuv test work

davidfowl committed:
Wait for the token to fire before disposing

davidfowl committed:
Transport is responsible for waiting on the token to fire
- We can now call DisposeAsync in the accept loop

davidfowl committed:
Wrote a test for unaccepted connections

davidfowl committed:
Added support for setting the read and write buffers on each transport

davidfowl committed:
Update the refs

davidfowl committed:
Remove scheduling mode

davidfowl commented on src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs:
```diff
@@ -234,7 +234,7 @@ private async Task StartAsyncCore(TransferFormat transferFormat)
         /// a new instance should be created using the same options.
         /// </remarks>
-        public async Task DisposeAsync()
+        public override async ValueTask DisposeAsync()
```
cc @BrennanConroy @anurse

davidfowl committed:
Fixed fallout from adding DisposeAsync to ConnectionContext

davidfowl committed:
Generated ref for connection abstractions

davidfowl committed:
Rename StopAsync to UnbindAsync and added ct to args

davidfowl committed:
Added some more LibuvTransportTests

davidfowl committed:
Moved TransportConnection to Connections.Abstractions

davidfowl committed:
Updated refs

davidfowl committed:
Moved FileHandleEndPoint to Connections.Abstractions

davidfowl committed:
Remove dependency on Transport.Abstractions

davidfowl committed:
Updated the refs

davidfowl commented:
/azp run

davidfowl commented:
AHHHHH

davidfowl committed:
Fixed breaks after rebase

davidfowl committed:
Handle a null pool

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs:
```diff
@@ -363,7 +362,7 @@ private static Pipe CreateDataPipe(MemoryPool<byte> pool)
                 resumeWriterThreshold: 1,
                 useSynchronizationContext: false,
-                minimumSegmentSize: KestrelMemoryPool.MinimumSegmentSize
+                minimumSegmentSize: pool.GetMinimumAllocSize()
```
This should be GetMinimumSegmentSize

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs:
```diff
@@ -494,7 +493,7 @@ private Pipe CreateRequestBodyPipe(uint windowSize)
                 resumeWriterThreshold: windowSize + 1,
                 useSynchronizationContext: false,
-                minimumSegmentSize: KestrelMemoryPool.MinimumSegmentSize
+                minimumSegmentSize: _context.MemoryPool.GetMinimumAllocSize()
```
This should be GetMinimumSegmentSize()

davidfowl committed:
Fixed segment size

davidfowl committed:
Updated the ref

davidfowl committed:
Use the right scheduler

davidfowl committed:
Skip ParseAddressUnixPipe on windows 7

davidfowl committed:
Only run unix pipe test on windows10

davidfowl commented on src/Servers/Kestrel/Core/test/AddressBinderTests.cs:
```diff
@@ -65,21 +65,20 @@ public void ParseAddressDefaultsToAnyIPOnInvalidIPAddress(string host)
         }
 
-        [Fact]
+        [OSSkipCondition(OperatingSystems.Windows, WindowsVersions.Win7, WindowsVersions.Win8, WindowsVersions.Win81, WindowsVersions.Win2008R2, SkipReason = "UnixDomainSocketEndPoint is not supported on older versions of Windows")]
```
This is breaking change as well, you can no longer construct a `UnixDomainSocketEndPoint` on an OS that doesn't support it.

jkotalik replied:
This is a rare case, correct?

davidfowl replied:
Correct, it would have failed later anyways.

davidfowl committed:
Don't capture the sync context

jkotalik commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelConnection.cs:
```diff
@@ -1,25 +1,168 @@
+
+        void IConnectionCompleteFeature.OnCompleted(Func<object, Task> callback, object state)
+        {
+            if (_completed)
```
Don't we have this same pattern with Request OnCompleted? Can we simplify/consolidate this logic somehow?

davidfowl replied:
This isn’t new code. It’s just moved, and it’s the same logic. If you have comments on what you’d like to simplify, it can be applied to both patterns in a later commit

jkotalik commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelConnection.cs:
```diff
@@ -1,25 +1,168 @@
+            }
+            catch (ObjectDisposedException)
+            {
+                // There's a race where the token could be disposed
```
Why/Where is there a race?

davidfowl replied:
Originally it was because the token was disposed in a different place than it was consumed. We can do this without the catch now.

davidfowl replied:
I take that back, the race is between shutting down and the connection ending. A call to RequestClose can be made “at the same time” a connection is being disposed.

jkotalik commented on src/Servers/Kestrel/Core/src/Internal/MemoryPoolExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+        public static int GetMinimumAllocSize(this MemoryPool<byte> pool)
+        {
+            // 1/2 of a segment
+            return pool.GetMinimumSegmentSize() / 2;
```
Wasn't this 1/4th before? Why change it?

davidfowl replied:
Because it’s 1/2 at the transport level I don’t know why it was 1/4 here. I made them all the same

jkotalik commented on src/Servers/Kestrel/Core/src/KestrelServer.cs:
```diff
@@ -79,27 +77,11 @@ private static ServiceContext CreateServiceContext(IOptions<KestrelServerOptions
                 Log = trace,
                 HttpParser = new HttpParser<Http1ParsingHandler>(trace.IsEnabled(LogLevel.Information)),
-                Scheduler = scheduler,
+                Scheduler = PipeScheduler.ThreadPool,
```
You said:
> We've removed the ability to set the scheduling mode on Kestrel, this will be a transport knob instead 

Where is this set?

davidfowl replied:
It’s not exposed in this PR. I’m not convinced we should yet.

jkotalik commented on src/Servers/Kestrel/Core/src/KestrelServer.cs:
```diff
@@ -182,8 +167,10 @@ public async Task StopAsync(CancellationToken cancellationToken)
                 {
-                    tasks[i] = _transports[i].UnbindAsync();
+                    (IConnectionListener listener, Task acceptLoop) = _transports[i];
+                    tasks[i] = Task.WhenAll(listener.UnbindAsync(cancellationToken).AsTask(), acceptLoop);
```
I'm not a fan of having to do Task.WhenAll here between unbinding and the accept loop. Can you clarify why you need to wait on both? How would someone else using these interfaces handle this?

davidfowl replied:
To dispose, we need to first unbind so no new connections can enter the accept loop, then we need to await the accept loop so that it drains existing connections before disposing.

jkotalik commented on src/Servers/Kestrel/Transport.Abstractions/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.netcoreapp3.0.cs:
```diff
@@ -1,113 +1,3 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.Internal
```
I didn't realize you already deleted all of this already. How much more work is it to remove it entirely?

davidfowl replied:
I was going to remove it but it’s part of every solution because we have all of the dependencies there. I can do it less properly and just remove it from this solution (my guess is that nobody else did this yet?)

jkotalik commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs:
```diff
@@ -31,31 +30,54 @@ internal partial class LibuvConnection : TransportConnection, IDisposable
         public override MemoryPool<byte> MemoryPool => Thread.MemoryPool;
-        public override PipeScheduler InputWriterScheduler => Thread;
-        public override PipeScheduler OutputReaderScheduler => Thread;
+        public void Start()
```
nit space.

jkotalik reviewed (commented):
Got through half of it. Will add some more feedback on the transports later.

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/ConnectionDispatcher.cs:
```diff
@@ -28,26 +26,46 @@ public ConnectionDispatcher(ServiceContext serviceContext, ConnectionDelegate co
+        private void StartAcceptingConnectionsCore(IConnectionListener listener)
+        {
+            // REVIEW: Multiple accept loops in parallel?
+            _ = AcceptConnectionsAsync();
```
Throughput. It’s not explicitly required unless we see a connection ramp up throughput problem

halter73 replied:
I'd say we should leave it to the transports to do the right thing. If having multiple loops is helpful, a transport can do that itself and merge the results into a single channel or something like libuv.

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/ConnectionDispatcher.cs:
```diff
@@ -28,26 +26,46 @@ public ConnectionDispatcher(ServiceContext serviceContext, ConnectionDelegate co
+                }
+                catch (Exception)
+                {
+                    // REVIEW: Today the only way to end the accept loop is an exception
```
Yes I forgot to log here and update the comment

davidfowl commented on src/Servers/Connections.Abstractions/src/Features/IConnectionEndpointFeature.cs:
```diff
@@ -0,0 +1,10 @@
+using System.Net;
```
Yea I always forget these

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/ConnectionDispatcher.cs:
```diff
@@ -28,26 +26,46 @@ public ConnectionDispatcher(ServiceContext serviceContext, ConnectionDelegate co
+                {
+                    while (true)
+                    {
+                        var connection = await listener.AcceptAsync();
```
What’s being boxed? These are all reference types.

halter73 replied:
I'm assuming you mean captured by the local function. `this` is captured too. You could make the local function static and pass in both the listener an this as arguments, but this is only called once to start the accept loop.

davidfowl committed:
Some PR feedback
- Fixed test bug found in another PR
- Added copyright

davidfowl committed:
PR feedback and fixes
- Log a critical error if the call to AcceptAsync throws.
- Fix null ref when AcceptAsync was pending during call to UnbindAsync or DisposeAsync
- Fixed libuv test that could throw if the dispose happened before
the connection was estabilished

anurse commented:
This feels like it would be good to get in for preview 6. What kind of e2e verification do we need here? We want to land everything in a candidate build by EOD Friday.

davidfowl commented:
/azp run

davidfowl commented:
> This feels like it would be good to get in for preview 6 because it's breaking. What kind of e2e verification do we need here? We want to land everything in a candidate build by EOD Friday.


Yep. Not much should be affected up stack but there are some settings that no longer work that exists on `KestrelSeverOptions`

davidfowl committed:
Handle connection resets in accept

davidfowl committed:
Set NoDelay to true by default

halter73 commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/ListenerContext.cs:
```diff
@@ -2,46 +2,70 @@
 {
     internal class ListenerContext
     {
+        private readonly Channel<LibuvConnection> _acceptQueue = Channel.CreateUnbounded<LibuvConnection>();
```
Should this be single reader?

```suggestion
        private readonly Channel<LibuvConnection> _acceptQueue = Channel.CreateUnbounded<LibuvConnection>(new UnboundedChannelOptions { SingleReader = true });
```

halter73 commented:
> TransportConnection has been moved to Connection.Abstractions (we can decide if we need to consolidate with DefaultConnectionContext in a later PR)

It seems like you want to do this. It's probably a good idea. Do you want to file an issue?

halter73 commented:
> NoDelay is a noop, on ListenOptions, we should consider deprecating it.

Deprecate it.

davidfowl commented:
> Deprecate it.

I'm removing it, prior art is MVC which did a similar thing for options that no longer work.

halter73 commented on src/Servers/Kestrel/Transport.Libuv/test/LibuvConnectionTests.cs:
```diff
@@ -19,167 +19,153 @@ public class LibuvConnectionTests
 
                 }, null);
 
+                var connection = await listenerContext.AcceptAsync();
```
It's nice how simplified this is.

halter73 commented on src/Servers/Kestrel/Transport.Libuv/test/LibuvTransportTests.cs:
```diff
@@ -23,71 +23,142 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.Tests
         }
 
-        [Theory]
-        [MemberData(nameof(ConnectionAdapterData))]
```
Why do you hate my variations‽ /s

davidfowl committed:
Update the ref assembly

halter73 commented on src/Servers/Kestrel/Transport.Libuv/test/ListenerPrimaryTests.cs:
```diff
@@ -270,44 +237,73 @@ public async Task PipeConnectionsWithWrongMessageAreLoggedAndIgnored()
-            }
+                using var socket = await HttpClientSlim.GetSocket(address);
+
+                await using var connection = await expected.DefaultTimeout();
```
Yeah, you did take some artistic license on this test. It would be nice if there was a quicker way we could verify HttpClientSlim actually has the same socket listenerPrimary does. Before, the "Primary" vs. "Secondary" responses gave you a quick error instead of this test failing with a timeout.

davidfowl replied:
It can't though. We need to verify what's in the queues, there's no way to have different "applications" for different listeners, the fact that it worked that way before was a bit of a fluke because we had a different dispatcher per listener and we took advantage of that.

halter73 replied:
You could write "Primary" or "Secondary" to the LibuvConnection after accepting it, no?

davidfowl replied:
Sure, I can do that but the real test is the fact that it actually is en-queued in the right queue, not the writing to the queue.

halter73 commented on src/Servers/Connections.Abstractions/ref/Microsoft.AspNetCore.Connections.Abstractions.netstandard2.0.cs:
```diff
@@ -30,12 +30,16 @@ public static partial class ConnectionBuilderExtensions
     public abstract partial class ConnectionContext
```
File an issue to make IDisposable.

davidfowl replied:
https://github.com/aspnet/AspNetCore/issues/10607

halter73 commented on src/Servers/Connections.Abstractions/ref/Microsoft.AspNetCore.Connections.Abstractions.netstandard2.0.cs:
```diff
@@ -70,40 +74,97 @@ public partial class ConnectionResetException : System.IO.IOException
         public ConnectionResetException(string message, System.Exception inner) { }
     }
-    public partial class DefaultConnectionContext : Microsoft.AspNetCore.Connections.ConnectionContext, Microsoft.AspNetCore.Connections.Features.IConnectionIdFeature, Microsoft.AspNetCore.Connections.Features.IConnectionItemsFeature, Microsoft.AspNetCore.Connections.Features.IConnectionLifetimeFeature, Microsoft.AspNetCore.Connections.Features.IConnectionTransportFeature, Microsoft.AspNetCore.Co ...
+    public partial class DefaultConnectionContext : Microsoft.AspNetCore.Connections.ConnectionContext, Microsoft.AspNetCore.Connections.Features.IConnectionEndPointFeature, Microsoft.AspNetCore.Connections.Features.IConnectionIdFeature, Microsoft.AspNetCore.Connections.Features.IConnectionItemsFeature, Microsoft.AspNetCore.Connections.Features.IConnectionLifetimeFeature, Microsoft.AspNetCore.Con ...
```
Should we remove the synchronous IDisposable from this?

davidfowl replied:
I'll file it under the same issue as IAsyncDisposable, I think that'll break some signalr tests.

halter73 commented on src/Servers/Connections.Abstractions/src/IConnectionListenerFactory.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Microsoft.AspNetCore.Connections
+{
+    public interface IConnectionListenerFactory
```
How nice would it be to avoid the word factory? Maybe IEndpointBinder?

davidfowl replied:
Not changing the name in this PR, we agreed to this one in the meeting, we can iterate on the issue.

davidfowl replied:
That said, I don't disagree

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -17,6 +17,18 @@ public class SocketTransportOptions
+
+        public long? MaxWriteBufferSize { get; set; } = PipeOptions.Default.PauseWriterThreshold;
+
+        internal Func<MemoryPool<byte>> MemoryPoolFactory { get; set; } = System.Buffers.MemoryPoolFactory.Create;
```
How many people do you expect to look at this and think "Oh look, *that's* where they wire up Kestrel's SlabMemoryPool"?

davidfowl replied:
Maybe like 4 people, 5 if we're lucky 😄

halter73 commented on src/Shared/Buffers.MemoryPool/MemoryPoolFactory.cs:
```diff
@@ -1,12 +1,9 @@
+namespace System.Buffers
 {
-    public static class KestrelMemoryPool
+    internal static class MemoryPoolFactory
```
Can we give this abetter name and/or namespace?

davidfowl replied:
Suggest a name, this is basically unnecessary churn at this point.

halter73 replied:
KestrelMemoryPoolFactory

halter73 replied:
in the old namespace

davidfowl replied:
Nah

halter73 replied:
Do you want it approved or not? 😉

halter73 commented on src/SignalR/SignalR.sln:
```diff
@@ -147,6 +147,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.HttpOv
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.SignalR.Protocols.Json", "common\Protocols.Json\src\Microsoft.AspNetCore.SignalR.Protocols.Json.csproj", "{BB52C0FB-19FD-485A-9EBD-3FC173ECAEA0}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Metadata", "..\Http\Metadata\src\Microsoft.AspNetCore.Metadata.csproj", "{2E107FBB-2387-4A9F-A2CA-EFDF2E4DD64D}"
```
You're adding IAllowAnonymous and IAuthorizeData to the SignalR solution in this PR?

davidfowl replied:
I needed it to compile, so pretend its ok. We added a new package but didn't add it to all the places. I had to make sure SignalR was working so I included this in the PR

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs:
```diff
@@ -32,44 +30,61 @@ internal sealed class SocketConnection : TransportConnection, IDisposable
         private volatile Exception _shutdownReason;
-
-        internal SocketConnection(Socket socket, MemoryPool<byte> memoryPool, PipeScheduler scheduler, ISocketsTrace trace)
+        private Task _task;
```
Come up with a better name please.

halter73 commented on src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs:
```diff
@@ -123,19 +161,25 @@ public async Task Start()
+            }
+
             _connectionClosedTokenSource.Dispose();
-            _connectionClosingCts.Dispose();
```
Why did this get removed?

davidfowl replied:
The transport no longer owns this but it's not being disposed right now. It's in KestrelConnection.

davidfowl replied:
Fixed

davidfowl commented:
/azp run

davidfowl committed:
PR feedback

davidfowl committed:
Handle connection reset

halter73 reviewed (approved):
Approved pending finding and fixing the root cause of the UvLoopHandle finalized. It would also be nice if you wrote "Primary"/"Secondary" to the accepted pipes in ListenerPrimaryTests and verified the response.

davidfowl committed:
Some cleanup
- Added ConfigureAwait(false) to DisposeAsync path
- Added await using in tests to make sure threads are disposed
- Remove more dead code

davidfowl committed:
Make everything async and clean up the accept loop

davidfowl commented:
> Approved pending finding and fixing the root cause of the UvLoopHandle finalized.

OK I spent a few hours on this and I'm still stumped and haven't been able to write a test to cause the condition to happen. I make a few tweaks to use await using in a couple of places just to make sure we're calling dispose in the right spots.

What if we just add a log [here](https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvThread.cs#L303) so that we can know if this is a resource leak or a bad unit test?

davidfowl committed:
Improve aborting queued connection logic
- Add a method to quickly run through each of the queued aborting all connections instead of using the IAsyncEnumerator.

davidfowl commented:
/ask run

davidfowl commented:
/azp run

halter73 commented:
> 
What if we just add a log here  so that we can know if this is a resource leak or a bad unit test?

Seems reasonable. We should probably log something anytime OnStopRude() let alone OnStopImmediate() is called. Maybe we could make those logs fail tests like the ungraceful shutdown log does.

jkotalik commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -17,6 +17,18 @@ public class SocketTransportOptions
+        /// </remarks>
+        public bool NoDelay { get; set; } = true;
+
+        public long? MaxReadBufferSize { get; set; } = PipeOptions.Default.PauseWriterThreshold;
```
Nit: doc comments?

davidfowl committed:
Added logs

davidfowl committed:
Log critical for ungraceful loop shutdown
